### PR TITLE
feat(android): add AndroidDevice.close event

### PR DESCRIPTION
**Pull request by mxschmitt**

### CODE REVIEW

File: tests/android/launch-server.spec.ts [hunk 1]
```diff
@@ -20,26 +20,34 @@ import { androidTest as test, expect } from './androidTest';
 test('android.launchServer should connect to a device', async ({ playwright }) => {
   const browserServer = await playwright._android.launchServer();
   const device = await playwright._android.connect(browserServer.wsEndpoint());
+  const closeEvents = [];
+  browserServer.on('close', () => closeEvents.push('close'));
   const output = await device.shell('echo 123');
   expect(output.toString()).toBe('123\n');
   await device.close();
+  expect(closeEvents).toEqual(['close']);
```
Comment (dgozman) [inline tests/android/launch-server.spec.ts:28]: Why closing the connected device closes the server? Shouldn't I be able to connect to it again?

File: packages/playwright-core/src/server/android/backendAdb.ts [hunk 2]
```diff
@@ -71,23 +71,26 @@ class AdbDevice implements DeviceBackend {
...
+    let commandOutput: Buffer;
+    if (!command.startsWith('shell:')) {
+      const remainingLength = parseInt((await socket.read(4)).toString(), 16);
+      commandOutput = await socket.read(remainingLength);
+    } else {
+      commandOutput = await socket.readAll();
+    }
+    return commandOutput;
+  } finally {
+    socket.close();
```
Comment (mxschmitt) [inline packages/playwright-core/src/server/android/backendAdb.ts:92]: drive-by: always close the socket.

File: tests/android/launch-server.spec.ts [hunk 3]
```diff
@@ -94,10 +112,12 @@ test('android.launchServer should terminate WS connection when device gets disco
   forwardingServer.on('connection', connection => {
     receivedConnection = connection;
     const actualConnection = new ws.WebSocket(browserServer.wsEndpoint());
-    actualConnection.on('message', message => connection.send(message));
```
Comment (mxschmitt) [inline tests/android/launch-server.spec.ts:97]: drive-by: unflake it by waiting for connection open.

File: packages/playwright-core/src/server/android/android.ts [hunk 4]
```diff
@@ -223,25 +229,29 @@ export class AndroidDevice extends SdkObject {
...
+    if (!driver)
+      throw new Error('Device is closed');
     const id = ++this._lastId;
     const result = new Promise((fulfill, reject) => this._callbacks.set(id, { fulfill, reject }));
     driver.send(JSON.stringify({ id, method, params }));
     return result;
   }
 
   async close() {
+    if (this._isClosed)
```
Comment (dgozman) [inline packages/playwright-core/src/server/android/android.ts:241]: Maybe we don't need this one?

File: packages/playwright-core/src/client/android.ts [hunk 5]
```diff
@@ -238,12 +240,16 @@ export class AndroidDevice extends ChannelOwner<channels.AndroidDeviceChannel> i
   }
 
   async close() {
+    if (this._closed)
```
Comment (dgozman) [inline packages/playwright-core/src/client/android.ts:243]: It seems like we don't need this one at all, looking at `browser.ts`.

File: packages/playwright-core/src/client/android.ts [hunk 6]
```diff
@@ -238,12 +240,16 @@ export class AndroidDevice extends ChannelOwner<channels.AndroidDeviceChannel> i
...
+      return;
+    this._closed = true;
     try {
-      this._didClose();
-      if (this._shouldCloseConnectionOnClose)
+      if (this._shouldCloseConnectionOnClose) {
         this._connection.close(kBrowserClosedError);
-      else
+        this.emit(Events.AndroidDevice.Close, this);
```
Comment (dgozman) [inline packages/playwright-core/src/client/android.ts:249]: `_didClose()` already does it.

**Review by dgozman — APPROVED** [review]
